### PR TITLE
feat(adapters): cloudflare-d1 - `mapRow` 

### PR DESCRIPTION
**Pull request by Druue**
contributes https://github.com/prisma/team-orm/issues/830

To be reviewed with https://github.com/prisma/prisma/pull/22794 which I merged but now needs a review as well.

To review D1 code globally: you can check https://github.com/prisma/prisma/tree/integration/d1-mapRow/packages/adapter-d1

### CODE REVIEW

File: packages/adapter-d1/src/conversion.ts [hunk 1]
```diff
@@ -75,48 +76,40 @@ class UnexpectedTypeError extends Error {
...
-//   }
-
-//   return result
-// }
+// TODO(@druue) This currently mimics mapD1ToRows without the extra type functionality.
+// TODO(@druue) next step is to implement further types.
+export function mapRow(obj: Object, _columnTypes: ColumnType[]): unknown[] {
+  const result: unknown[] = Object.keys(obj).map((k) => obj[k])
+
+  console.log(result)
```
Comment (Jolg42) [inline packages/adapter-d1/src/conversion.ts:84]: To be removed later

File: packages/adapter-d1/src/conversion.ts [hunk 2]
```diff
@@ -8,6 +8,8 @@ import {
 
 export type Value = null | string | number | object
 
+export const absoluteBigint = (n) => (n < 0n ? -n : n)
```
Comment (Jolg42) [inline packages/adapter-d1/src/conversion.ts:11]: Not used at the moment, I thought it was good to keep in case we decide to do something with it

File: packages/client/tests/functional/raw-queries/typed-results/tests.ts [hunk 3]
```diff
@@ -28,37 +39,198 @@ testMatrix.setupTestSuite(
...
       }
     })
   },
   {
     optOut: {
-      from: ['mongodb', 'mysql'],
+      from: ['mongodb'],
       reason: `
         $queryRaw only works on SQL based providers
-        mySql You have an error in your SQL syntax; check the manual that corresponds to your MariaDB server version for the right syntax to use near '"TestModel"'
```
Comment (Jolg42) [inline packages/client/tests/functional/raw-queries/typed-results/tests.ts:61]: I made this test run on MySQL, it was easy to do

File: packages/client/tests/functional/raw-queries/typed-results/tests.ts [hunk 4]
```diff
@@ -28,37 +39,198 @@ testMatrix.setupTestSuite(
...
+          //   allocUnsafeSlow: [Function: allocUnsafeSlow],
+          //   isBuffer: [Function: isBuffer],
+          //   compare: [Function: compare],
+          //   isEncoding: [Function: isEncoding],
+          //   concat: [Function: concat],
+          //   byteLength: [Function: byteLength],
+          //   [Symbol(kIsEncodingSymbol)]: [Function: isEncoding]
+          // }
+          bytes: clientRuntime === 'wasm' ? expect.anything() : Buffer.from([1, 2, 3]),
+          bool: driverAdapter === 'js_d1' || provider === 'mysql' ? 1 : true,
```
Comment (Jolg42) [inline packages/client/tests/functional/raw-queries/typed-results/tests.ts:86]: It's expected for D1 but I didn't expect this for MySQL / PlanetScale 🤔 Though I didn't check more than adjusting the test after I made it runnable on MySQL

File: packages/client/tests/functional/raw-queries/typed-results/tests.ts [hunk 5]
```diff
@@ -6,17 +7,27 @@ declare let prisma: PrismaClient
...
+    }
+
     test('simple expression', async () => {
       const result = (await prisma.$queryRaw`SELECT 1 + 1`) as Array<Record<string, unknown>>
       expect(Number(Object.values(result[0])[0])).toEqual(2)
     })
 
-    // TODO: The buffer binary data does not match the expected one
-    skipTestIf(clientRuntime === 'wasm')('query model with multiple types', async () => {
+    test('query model with multiple types', async () => {
```
Comment (Jolg42) [inline packages/client/tests/functional/raw-queries/typed-results/tests.ts:28]: Instead of skipping the test for "wasm", I made it ignore the result for `bytes` for "wasm", see `expect.anything()` below

File: packages/client/tests/functional/raw-queries/typed-results/prisma/_schema.ts [hunk 6]
```diff
@@ -12,7 +13,7 @@ export default testMatrix.setupSchema(({ provider }) => {
     }
     
     model TestModel {
-      id     Int       @id
+      id    ${idForProvider(provider)}
```
Comment (Jolg42) [inline packages/client/tests/functional/raw-queries/typed-results/prisma/_schema.ts:16]: Since I added tests and I didn't want to make the tests rely on the id I tweaked this, it does not change the test, not in what's tested.

File: packages/adapter-d1/src/d1.ts [hunk 7]
```diff
@@ -58,73 +51,42 @@ class D1Queryable<ClientT extends StdClient> implements Queryable {
...
-
-        return arg
-      })
+      query.args = query.args.map((arg) => this.cleanArg(arg))
 
       const result = await this.client
         .prepare(query.sql)
         .bind(...query.args)
+        // TODO use .raw({ columnNames: true }) later
         .all()
```
Comment (Jolg42) [inline packages/adapter-d1/src/d1.ts:88]: We need to wait for a release from their side, though it's in the runtime already in the Cloud
and it will be best in a separate PR

File: packages/debug/src/index.ts [hunk 8]
```diff
@@ -156,6 +156,8 @@ function safeStringify(value: any, indent = 2) {
         }
 
         cache.add(value)
+      } else if (typeof value === 'bigint') {
+        return value.toString()
```
Comment (Jolg42) [inline packages/debug/src/index.ts:160]: Note for @millsp 
I think you're ok with that?

**Comment by Jolg42** [comment]
Note: the most recent integration version as of now if `5.10.0-integration-d1-map-row.11`

File: packages/adapter-d1/src/conversion.ts [hunk 9]
```diff
@@ -28,43 +30,70 @@ export function getColumnTypes(columnNames: string[], rows: Object[]): ColumnTyp
...
   }
 }
 
-function inferObjectType(value: {}): ColumnType {
+// See https://stackoverflow.com/a/3143231/1345244
+const isoDateRegex = new RegExp(
+  /(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d\.\d+([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))|(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d([+-][0-2]\d:[0-5]\d|Z))/,
+)
+function isISODate(str) {
+  return isoDateRegex.test(str)
```
Comment (SevInf) [inline packages/adapter-d1/src/conversion.ts:66]: I'd prefer `new Date(str) !== 'Invalid Date'`. It lets in a little bit more than only ISO dates, but no regex.

File: packages/adapter-d1/src/d1.ts [hunk 10]
```diff
@@ -149,6 +111,34 @@ class D1Queryable<ClientT extends StdClient> implements Queryable {
...
+    // * ✘ [ERROR] Error in performIO: Error: D1_TYPE_ERROR: Type 'boolean' not supported for value 'true'
+    if (arg === true) {
+      return 1
+    }
+
+    if (arg === false) {
+      return 0
+    }
+
+    if (arg instanceof Uint8Array) {
```
Comment (SevInf) [inline packages/adapter-d1/src/d1.ts:126]:
```suggestion
    if (ArrayBuffer.isView(arg)) {
```

File: packages/adapter-d1/src/conversion.ts (hunk 9 above)
Comment (Jolg42) [reply packages/adapter-d1/src/conversion.ts:66]: But we have no idea what's in the string, so it would result in weird things like for example
```js
new Date('Hello world 5')
-> Date Tue May 01 2001 00:00:00 GMT+0200 (Central European Summer Time)
```

File: packages/adapter-d1/src/conversion.ts [hunk 11]
```diff
@@ -80,48 +109,35 @@ class UnexpectedTypeError extends Error {
...
-//     if (typeof value === 'bigint') {
-//       result[i] = value.toString()
-//       continue
-//     }
-//   }
-
-//   return result
-// }
+export function mapRow(obj: Object, columnTypes: ColumnType[]): unknown[] {
+  const result: unknown[] = Object.keys(obj).map((k) => obj[k])
```
[the remark this answers is not in the document]
Comment (SevInf) [reply packages/adapter-d1/src/conversion.ts:113]: Also, i heard D1 now has a way to get a row as an array. If that is true, i'd suggest to switch to that method as there are some aggregation queries client produces that have several columns with identical names. Results can not be read from all of them using object mode.

File: packages/client/helpers/functional-test/run-tests.ts [hunk 12]
```diff
@@ -157,6 +158,18 @@ async function main(): Promise<number | void> {
...
+      // Because we cannot reset the database contents programmatically at the moment,
+      // deleting it is the easy way
+      // It makes local tests consistently fast and clean
+      try {
+        fs.rmdirSync(path.join(__dirname, '..', '..', '.wrangler'), { recursive: true })
+      } catch (e) {
+        if (e.code !== 'ENOENT') {
+          console.warn('Deleting the .wrangler directory failed with:', e)
+        }
+      }
```
Comment (SevInf) [inline packages/client/helpers/functional-test/run-tests.ts:171]:
```suggestion
      fs.rmSync(path.join(__dirname, '..', '..', '.wrangler'), { recursive: true, force: true })
```

`rm -rf` equivalent that does not errors on non existent dir.

File: packages/adapter-d1/src/conversion.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (Jolg42) [reply packages/adapter-d1/src/conversion.ts:113]: Yes, I think we should switch but like I mentioned in a comment in this PR, better to be done separately, also we need to wait for some releases to happen in the tooling first.

**Review by SevInf — APPROVED** [review]
Nothing blocking from my side, good job.

File: packages/adapter-d1/src/d1.ts (hunk 10 above)
Comment (Jolg42) [reply packages/adapter-d1/src/d1.ts]: I reverted this for now because I encountered this error
```
Error: packages/adapter-d1 build: src/d1.ts(127,25): error TS2769: No overload matches this call.
packages/adapter-d1 build:   Overload 1 of 4, '(iterable: Iterable<unknown> | ArrayLike<unknown>): unknown[]', gave the following error.
packages/adapter-d1 build:     Argument of type 'ArrayBufferView' is not assignable to parameter of type 'Iterable<unknown> | ArrayLike<unknown>'.
packages/adapter-d1 build:   Overload 2 of 4, '(arrayLike: ArrayLike<unknown>): unknown[]', gave the following error.
packages/adapter-d1 build:     Argument of type 'ArrayBufferView' is not assignable to parameter of type 'ArrayLike<unknown>'.
packages/adapter-d1 build:       Property 'length' is missing in type 'ArrayBufferView' but required in type 'ArrayLike<unknown>'.
Error: s/adapter-d1 build: ✘ [ERROR] Command failed with exit code 2: tsc --project tsconfig.build.json --paths null [plugin tscPlugin]
```
Which I first tried to solve by changing
```
if (ArrayBuffer.isView(arg)) {
      return Array.from(arg)
}
```
to
```js
if (ArrayBuffer.isView(arg)) {
    return Buffer.from(arg.buffer, arg.byteOffset, arg.byteLength)
}
```

Which in turn resulted in errors like 
```
 FAIL  tests/functional/raw-queries/send-type-hints/tests.ts
  ● Console

    console.error
      Error in performIO: Error [AssertionError]: false == true
          at new AssertionError (node-internal:internal_assertionerror:419:15)
          at assert (node-internal:internal_assert:49:15)
          at ArrayBufferView (file:///Users/j42/Dev/prisma/node_modules/.pnpm/miniflare@3.20240129.2/node_modules/miniflare/src/workers/core/devalue.ts:95:3)
          at hydrate (file:///Users/j42/Dev/prisma/node_modules/.pnpm/miniflare@3.20240129.2/node_modules/.pnpm/devalue@4.3.2/node_modules/devalue/src/parse.js:60:32)
          at hydrate (file:///Users/j42/Dev/prisma/node_modules/.pnpm/miniflare@3.20240129.2/node_modules/.pnpm/devalue@4.3.2/node_modules/devalue/src/parse.js:115:17)
          at unflatten (file:///Users/j42/Dev/prisma/node_modules/.pnpm/miniflare@3.20240129.2/node_modules/.pnpm/devalue@4.3.2/node_modules/devalue/src/parse.js:132:9)
          at parse (file:///Users/j42/Dev/prisma/node_modules/.pnpm/miniflare@3.20240129.2/node_modules/.pnpm/devalue@4.3.2/node_modules/devalue/src/parse.js:16:9)
          at parseWithReadableStreams (file:///Users/j42/Dev/prisma/node_modules/.pnpm/miniflare@3.20240129.2/node_modules/miniflare/src/workers/core/devalue.ts:313:9)
          at #fetch (file:///Users/j42/Dev/prisma/node_modules/.pnpm/miniflare@3.20240129.2/node_modules/miniflare/src/workers/core/proxy.worker.ts:237:12)
          at async ProxyServer.fetch (file:///Users/j42/Dev/prisma/node_modules/.pnpm/miniflare@3.20240129.2/node_modules/miniflare/src/workers/core/proxy.worker.ts:125:11) {
        [cause]: undefined
      }
```
